tempesta_fw/http_parser.c

```diff
@@ -2428,274 +2428,237 @@ __req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len)
+	__FSM_STATE(I_WDate4) {
+		if (likely(c == ',')) {
+			parser->date_fmt = &rfc822_fmt;
+			__FSM_I_MOVE(I_WDateSP);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date = __year_day_secs(parser->_acc, parser->_date);
-		if (parser->_date < 0)
-			return CSTR_NEQ;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateHourSP, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateHourSP) {
-		if (likely(c == ' '))
-			__FSM_I_MOVE(I_DateHour);
-		return CSTR_NEQ;
-	}
-
-	__FSM_STATE(I_DateHour) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_a(p, __fsm_sz, colon_a, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateHour;
-			__msg_hdr_chunk_fixup(data, len);
+		if ('a' <= c && c <= 'z') {
+			parser->date_fmt = &rfc850_fmt;
+			__FSM_I_MOVE(I_WDate5);
+		}
+		if (c == ' ') {
+			parser->date_fmt = &isoc_fmt;
+			__FSM_I_MOVE(I_Loop);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc * 3600;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateMinCln, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateMinCln) {
-		if (likely(c == ':'))
-			__FSM_I_MOVE(I_DateMin);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateMin) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_a(p, __fsm_sz, colon_a, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateMin;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_WDate5) {
+		if ('a' <= c && c <= 'z') {
+			__FSM_I_MOVE(I_WDate5);
+		}
+		if (c == ',') {
+			__FSM_I_MOVE(I_WDateSP);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc * 60;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateSecCln, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateSecCln) {
-		if (likely(c == ':'))
-			__FSM_I_MOVE(I_DateSec);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateSec) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_ws(p, __fsm_sz, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateSec;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_WDateSP) {
+		if (likely(c == ' ')) {
+			parser->_acc = 0;
+			__FSM_I_MOVE(I_Loop);
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateSecSP, __fsm_n);
-	}
-
-	__FSM_STATE(I_DateSecSP) {
-		if (likely(c == ' '))
-			__FSM_I_MOVE(I_DateZone);
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateZone) {
-		TRY_STR("gmt", I_DateZone, I_EoL);
-		TRY_STR_INIT();
-		return CSTR_NEQ;
+	__FSM_STATE(I_Loop) {
+		TfwHttpDateFmt *fmt = parser->date_fmt;
+
+		if (parser->_acc >= fmt->len) {
+			time_t date = __parse_date(parser->date_str, fmt);
+			if (date < 0)
+				return CSTR_NEQ;
+			parser->_date = date;
+			__FSM_I_JMP(I_EoL);
+		}
+		if (c < fmt->min[parser->_acc] || c > fmt->max[parser->_acc]) {
+			if (likely(fmt->type != ISOC ||
+			           parser->_acc != fmt->day_pos ||
+			           c != ' '))
+			{
+				return CSTR_NEQ;
+			}
+		}
+		parser->date_str[parser->_acc++] = c;
+		__FSM_I_MOVE(I_Loop);
```

Comment:
I don't like what happens in the state at all.

Please use bits [56-61] of `_acc` (bits [62-62] are used to index array of formats) for current parser position on the data string. If it reaches `fmt->len`, then return `CSTR_NEQ`. In `TfwHttpDateFmt` you have positions of the year, month, day and time, so by reaching each of the position we should move to appropriate states like `I_DateYear`, `I_DateMonth` etc. Your format templates also have separators, so we can use them in only one state for all the separators. States like `I_DateYear` just use `_acc` as previously to accumulate current integer values and do appropriate part of the logic from `__parse_date()`.

For me it looks like we can keep very similar code as the original `__parse_http_date()` with the extension by the format templates.